src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -702,5 +703,37 @@ private static string GetPropertyName(MemberInfo property)
 
             return property.Name;
         }
+
+        private static IEnumerable<IConfigurationSection> GetChildrenOrSelf(IConfiguration config)
+        {
+            if (!ShouldBindSingleElementsToArray())
+            {
+                return config.GetChildren();
+            }
+
+            IEnumerable<IConfigurationSection> children;
+            // If configuration's children is an array, the configuration key will be a number
+            if (config.GetChildren().Any(a => long.TryParse(a.Key, out _)))
+            {
+                children = config.GetChildren();
+            }
+            else
+            {
+                children = new[] { config as IConfigurationSection };
+            }
+
+            return children;
+        }
+
+        private static bool ShouldBindSingleElementsToArray()
+        {
+            if (AppContext.TryGetSwitch(BindSingleElementsToArraySwitch, out bool bindSingleElementsToArray))
```

Comment:
Does this need to be cached? Reading from AppContext each time in a loop might be a perf issue.